Hotfix/blas svd deflate

maddyscientist opened the pull request:
This is primarily a bugfix pull
* Fix SVD deflation what was broken by #855 
* Fix silent blas failure when using quarter-precision fields with float-2 ordering
* Add support for quarter precision to alternative reliable updates
* `EigenSolver` methods `loadVector` and `saveVector` should be static - fixes undefined behaviour from calling a method on a `nullptr`
* Robustness and cleanup of multi-blas framework.  Fixes an ASAN warning.

weinbe2 commented:
Not sure how to revert my approval, but for the time being don't merge this PR, I found a bug that @maddyscientist is going to hunt down. It can be reproduced by compiling with quarter precision support (in my case, I also had MPI on, and turned off domain wall just for compile time reasons; I imagine neither of these make a difference) and running `./blas_test` without any arguments.

maddyscientist commented:
Just pushed a commit to address @weinbe2's comments:
* Extended `blas_test` to include testing of float2-ordered nSpin-4 fields when `GPU_MULTIGRID` is enabled.  This fixes a hole in the prior coverage, and ensures that none of the regressions fixed in this branch can occur again.
* Fix multiple float2-ordered field blas-related bugs
* Default internal field order for nSpin-2 fields should be float2 ordering

weinbe2 commented:
Unless the other edits added coverage for this codepath in a way I didn't notice, this still needs the edit to line 222 of `color_spinor_param.h`:

` fieldOrder = (precision == QUDA_DOUBLE_PRECISION || nSpin == 1) ? QUDA_FLOAT2_FIELD_ORDER : QUDA_FLOAT4_FIELD_ORDER;`

Adding `||nSpin == 2`, or maybe by calling `setPrecision`?

maddyscientist commented:
@weinbe2 I've fixed color_spinor_field.h now: previously I'd only caught one of the constructors for `ColorSpinorParam`, but all good now.

@mathiaswagner I've added your patch to this branch to not include the cublas headers unless`CUBLAS_LIB` is enabled.  Amusingly, in doing so, I also found a bug (missing header in malloc_quda.h).

Conclusion: this pull is now complete, and can be merged when the build tests have completed.

weinbe2 commented:
@maddyscientist FYI I'm doing a quick build right now (which I imagine will finish before the build tests) to do some tests.

weinbe2 commented:
Running with `Nspin == 1`, (that is, `./blas_test --dslash-type asqtad`), getting the following two failures:

```
[ RUN      ] QUDA/BlasTest.verify/HeavyQuarkResidualNorm_single_default
/home/eweinberg/Data/quda_custom/pr894/quda/tests/blas_test.cu:1136: Failure
Expected: (deviation) <= (tol), actual: 2.25033 vs 1e-06
CPU and CUDA implementations do not agree
[  FAILED  ] QUDA/BlasTest.verify/HeavyQuarkResidualNorm_single_default, where GetParam() = (2, 29, 0) (528 ms)
[ RUN      ] QUDA/BlasTest.verify/xpyHeavyQuarkResidualNorm_single_default
/home/eweinberg/Data/quda_custom/pr894/quda/tests/blas_test.cu:1136: Failure
Expected: (deviation) <= (tol), actual: 2.25037 vs 1e-06
CPU and CUDA implementations do not agree
[  FAILED  ] QUDA/BlasTest.verify/xpyHeavyQuarkResidualNorm_single_default, where GetParam() = (2, 30, 0) (629 ms)
```

This is a run with `QUDA_PRECISION == 15`. I could reproduce the error by re-running the above command. To be explicit, the two blas routines do _not_ fail with quarter, half, or double.

weinbe2 commented:
Doing one last build!
